### ISSUE TITLE
The editor could not be opened because the file was not found.

### ISSUE THREAD

**Issue by davidlehrian** [microsoft/vscode-cpptools#10630]
### Environment

- OS and version: Windows 10
- VS Code: Version: 1.76.0 (system setup)
Commit: 92da9481c0904c6adfe372c12da3b7748d74bdcb
Date: 2023-03-01T10:22:44.506Z
Electron: 19.1.11
Chromium: 102.0.5005.196
Node.js: 16.14.2
V8: 10.2.154.26-electron.0
OS: Windows_NT x64 10.0.19045
Sandboxed: No
- C/C++ extension: v1.3.0
- OS and version of remote machine (if applicable):
- GDB / LLDB version: 11.2

### Bug Summary and Steps to Reproduce

Bug Summary:
GDB can't find source code for executable. 

Steps to reproduce:
1. create new vs code folder and add helloworld.cpp 
2. create build task using cygwin g++.exe
3. create launch task that references C:/cygwin/bin/gdb.exe
4. Set breakpoint in helloworld.cpp
5. Launch debugger and it doesn't stop at the breakpoint in the file.
6. set "stopAtEntry": true, in launch configuration and the error "The editor could not be opened because the file was not found." is displayed. This is how I figured out that it couldn't find my source code. 
7. The program runs just fine so the compiler is working, it just can't seem to find the source code.

### Debugger Configurations

```shell
{
    "configurations": [
    {
        "name": "(gdb) Launch",
        "type": "cppdbg",
        "request": "launch",
        "program": "${workspaceFolder}/helloworld.exe",
        "args": [],
        "stopAtEntry": true,
        "logging": {
            "engineLogging": true
         },
        "cwd": "${fileDirname}",
        "environment": [],
        "externalConsole": false,
        "MIMode": "gdb",
        "miDebuggerPath": "C:/cygwin/bin/gdb.exe",
        "setupCommands": [
            {
                "description": "Enable pretty-printing for gdb",
                "text": "-enable-pretty-printing",
                "ignoreFailures": true
            },
            {
                "description": "Set Disassembly Flavor to Intel",
                "text": "-gdb-set disassembly-flavor intel",
                "ignoreFailures": true
            }
        ],
        "preLaunchTask": "C/C++: g++.exe build active file"
    }
    ]
}
```

### Debugger Logs

So the error appears to be caused by the following line:

```shell
1: (1059) ->~"Thread 1 \"helloworld\" hit Breakpoint 1, main () at C:\\Users\\lehrian\\Documents\\CProjects\\HW\\helloworld.cpp:9\n"
Thread 1 "helloworld" hit Breakpoint 1, main () at C:\Users\lehrian\Documents\CProjects\HW\helloworld.cpp:9
1: (1062) ->&"9\tC:\\Users\\lehrian\\Documents\\CProjects\\HW\\helloworld.cpp: No such file or directory.\n"
1: (1062) ->*stopped,reason="breakpoint-hit",disp="keep",bkptno="1",frame={addr="0x0000000100401098",func="main",args=[],file="C:\\Users\\lehrian\\Documents\\CProjects\\HW\\helloworld.cpp",fullname="/cygdrive/c/Users/lehrian/Documents/CProjects/HW/C:\\Users\\lehrian\\Documents\\CProjects\\HW\\helloworld.cpp",line="9",arch="i386:x86-64"},thread-id="1",stopped-threads="all"
```

It finds it's way to the correct folder "/cygdrive/c/Users/lehrian/Documents/CProjects/HW/" but then instead of just appending the file name that I launched it appends the full path to the file that I launched "C:\\Users\\lehrian\\Documents\\CProjects\\HW\\helloworld.cpp" which doesn't work. 

I have hacked on this to try and use sourceFileMap to map "C:\\Users\\lehrian\\Documents\\CProjects\\HW\\" to "/" and '.' and '' but nothing seems to work. I can't seem to figure out how to tell it to not prepend the full path to the source file name. Any help would be greatly appreciated.

**Comment by sean-mcmanus** [comment]
I think you need to build with -g in args, i.e.
```json
            "args": [
                "-fdiagnostics-color=always",
                "-g",
                "${file}",
                "-o",
                "${fileDirname}\\${fileBasenameNoExtension}.exe"
            ],
```

**Comment by davidlehrian** [comment]
Thanks for the reply Sean, but I don't have a problem building the executable. And the executable runs fine from the command line. The issue is the debugger not finding the source code because it appears that somewhere in the configuration the debugger is being told/is assuming file="C:\\Users\\lehrian\\Documents\\CProjects\\HW\\helloworld.cpp" and not just file="helloworld.cpp". This is causing the debugger to look for the source file at fullname="/cygdrive/c/Users/lehrian/Documents/CProjects/HW/C:\\Users\\lehrian\\Documents\\CProjects\\HW\\helloworld.cpp" when it should be looking at fullname="/cygdrive/c/Users/lehrian/Documents/CProjects/HW/helloworld.cpp". I'm not sure where this is set or configured or if it is possibly a bug cpp-tools. 

But for completeness here is my build script which does contain the "-g". 
```
{
    "tasks": [
        {
            "type": "cppbuild",
            "label": "C/C++: g++.exe build active file",
            "command": "C:/cygwin/bin/g++.exe",
            "args": [
                "-fdiagnostics-color=always",
                "-g",
                "${file}",
                "-o",
                "${fileDirname}\\${fileBasenameNoExtension}.exe"
            ],
            "options": {
                "cwd": "${fileDirname}"
            },
            "problemMatcher": [
                "$gcc"
            ],
            "group": {
                "kind": "build",
                "isDefault": true
            },
            "detail": "Task generated by Debugger."
        }
    ],
    "version": "2.0.0"
}
```

**Comment by WardenGnaw** [comment]
Looks like the cygwin detection is not detecting correctly and not prefexing the path correctly.

https://github.com/microsoft/MIEngine/blob/153a9e56d7faf59f14f03c47234f7945d803b2fa/src/MIDebugEngine/Engine.Impl/CygwinFileMapper.cs#L69

**Comment by rafa-leo** [comment]
@davidlehrian, 

I faced the same problem, there is a workaround  found : [here
](https://stackoverflow.com/questions/58678800/how-to-fix-file-not-found-error-in-debugging-with-vscode-debugger-gcc
)
Adaptation which worked for me: 
Change your tasks.json from `"args":[ "${file}" ]` to `"args":[ "${relativeFile}"]` and change "cwd" to `${workspaceFolder}`.

It is not a perfect solution because the breakpoints still do not work, I have to start debugging from entry point. But still is better than before.

**Comment by davidlehrian** [comment]
Thanks for the workaround. I actually found that all that is needed is to change `"args":[ "${file}" ] `to `"args":[ "${relativeFile}"]` in tasks.json and it now finds my source code. It also stops at the breakpoints. 

But then I added a class file to my project and it no longer worked. It did not compile and link my class. So after a little searching I found that instead of `"args":[ "${relativeFile}"]` I could change it to `"args":[ "*.cpp"]` and it compiled and linked both files and I could set breakpoints in both main and my class and debug everything fine. 

However, in searching for a solution to this I did come across the suggestion to use Make or CMake build system as a better solution so I will be looking it that as I use CMake in my C++ projects in Eclipse. But for a simple project this certainly works. Thanks for your input!
